### ISSUE TITLE
Flash Attention causal kernel produces incorrect output (masked by overly loose test tolerances)

### ISSUE THREAD

**Issue by tonyjie** [Xilinx/mlir-air#1427]
# Flash Attention causal kernel produces incorrect output (masked by overly loose test tolerances)

## Summary

The `flash_attention/kernel_fusion_based` causal kernel (`--causal` flag) produces output with **correlation 0.086** against its own NumPy reference (`flash_attn_per_stage`), yet the standalone test reports `PASS!`. The test passes because the combination of `atol=0.5, rtol=0.2` with narrow-range test data `[0.5, 1.0]` creates a tolerance window so wide that any output in `[-0.2, 1.7]` is accepted — regardless of correctness.

This bug blocks LLAMA-3.2-1B inference: the attention kernel is the only incorrect step (all other 14 kernels have corr>0.999), and the error compounds across 16 transformer layers to produce meaningless output.

## Environment

- Hardware: AMD NPU2 (Strix, AIE2P)
- mlir-air: current main branch (post causal BD-exhaustion fix)
- Config: LQ=2048, LK=2048, LQP=256, LKP=64, DK=64, DV=64, NUM_HEADS=32, NUM_KV_HEADS=8

### 1. Current test passes (false positive)

```bash
cd programming_examples/flash_attention/kernel_fusion_based
make run LQ=2048 LK=2048 LQP=256 LKP=64 DK=64 DV=64 \
  NUM_HEADS=32 NUM_KV_HEADS=8 EXTRA_PY_FLAGS="--causal"
# Output: PASS!
```

### 2. Verify the output is actually wrong

Save the reproducer script below as `repro.py`, then run:

```bash
cd programming_examples/flash_attention/kernel_fusion_based/build_peano
python3 repro.py
```

Expected output:
```
NPU vs reference:  max_err=0.4922, corr=0.086 (essentially uncorrelated)

Tolerance sweep:
  atol=0.5, rtol=0.2 (current)    PASS  (0.00% elements fail)  ← false positive
  atol=0.1, rtol=0.1              FAIL  (0.06% elements fail)
  atol=0.05, rtol=0.05            FAIL  (0.34% elements fail)
  atol=0.01, rtol=0.01            FAIL  (10.72% elements fail)
```

<details>
<summary>Reproducer script (repro.py) — click to expand</summary>

```python
#!/usr/bin/env python3
"""Reproducer: causal flash attention kernel produces incorrect output.

Demonstrates that the causal kernel output has near-zero correlation with
its own NumPy reference, but the standalone test reports PASS due to
overly loose tolerances (atol=0.5, rtol=0.2) combined with narrow-range
test data ([0.5, 1.0]).

Usage:
    cd programming_examples/flash_attention/kernel_fusion_based/build_peano
    python3 repro.py

Prerequisites:
    Causal kernel must be compiled first:
    make run LQ=2048 LK=2048 LQP=256 LKP=64 DK=64 DV=64 \
      NUM_HEADS=32 NUM_KV_HEADS=8 EXTRA_PY_FLAGS="--causal"
"""

import numpy as np
from ml_dtypes import bfloat16
from math import sqrt
from pathlib import Path
from air.backend.xrt import XRTBackend, XRTCompileArtifact
import filelock

# === Parameters (LLAMA-3.2-1B, seq_len=2048) ===
LQ = LK = 2048
LKP, LQP, DK, DV = 64, 256, 64, 64
NUM_HEADS, NUM_KV_HEADS = 32, 8
NUM_CASCADE_STAGES, NUM_Q_TILES = 4, 4
GQA_GROUP_SIZE = NUM_HEADS // NUM_KV_HEADS
CHUNKS_PER_STAGE = (LK // LKP) // NUM_CASCADE_STAGES
BF16_LOWEST = np.float32(
    np.frombuffer(np.array([0xFF7F], dtype=np.uint16).tobytes(), dtype=bfloat16)[0]
)

# === Generate test data (SAME as standalone attn.py) ===
rng = np.random.default_rng(42)
input_q = (rng.uniform(0, 1, (NUM_HEADS, LQ, DK)) * 0.5 + 0.5).astype(bfloat16)
input_k = (rng.uniform(0, 1, (NUM_KV_HEADS, LK, DK)) * 0.5 + 0.5).astype(bfloat16)
input_v = (rng.uniform(0, 1, (NUM_KV_HEADS, LK, DV)) * 0.5 + 0.5).astype(bfloat16)
input_m = np.zeros((NUM_HEADS, LQ, LK), dtype=bfloat16)
scale = 1.0 / sqrt(DK)
input_q_scaled = (input_q.astype(np.float32) * scale).astype(bfloat16)

# === Compute reference (SAME as attn.py's flash_attn_per_stage) ===
print("Computing reference...")
input_k_f32 = input_k.astype(np.float32)
input_v_f32 = input_v.astype(np.float32)

def flash_attn_per_stage(A, kv_h, stage, mask_h):
    Gp = np.zeros((LQ, DV), dtype=np.float32)
    up = np.full((LQ, 1), BF16_LOWEST, dtype=np.float32)
    sp = np.zeros((LQ, 1), dtype=np.float32)
    for ci in range(CHUNKS_PER_STAGE):
        j = stage * CHUNKS_PER_STAGE + ci
        G = mask_h[:, j*LKP:(j+1)*LKP].astype(np.float32)
        B = input_k_f32[kv_h, j*LKP:(j+1)*LKP, :].T
        G = A.astype(np.float32) @ B + G
        for qi in range(LQ):
            for ki in range(LKP):
                if j*LKP + ki > qi:
                    G[qi, ki] = BF16_LOWEST
        u = np.maximum(np.max(G, axis=1, keepdims=True), BF16_LOWEST)
        u_new = np.maximum(up, u)
        G = np.exp(G - u_new)
        r = np.exp(up - u_new)
        Gp = Gp * r + G @ input_v_f32[kv_h, j*LKP:(j+1)*LKP, :]
        sp = sp * r + np.sum(G, axis=1, keepdims=True)
        up = u_new
    return Gp, up, sp

def cascade_merge(Gp_A, up_A, sp_A, Gp_B, up_B, sp_B):
    u_new = np.maximum(up_A, up_B)
    r_A, r_B = np.exp(up_A - u_new), np.exp(up_B - u_new)
    return Gp_A*r_A + Gp_B*r_B, u_new, sp_A*r_A + sp_B*r_B

ref_output = np.empty((NUM_HEADS, LQ, DV), dtype=bfloat16)
for h in range(NUM_HEADS):
    kv_h = h // GQA_GROUP_SIZE
    stages = [flash_attn_per_stage(input_q_scaled[h], kv_h, s, input_m[h])
              for s in range(NUM_CASCADE_STAGES)]
    Gp, up, sp = stages[-1]
    for s in range(NUM_CASCADE_STAGES-2, -1, -1):
        Gp, up, sp = cascade_merge(Gp, up, sp, *stages[s])
    ref_output[h] = (Gp / sp).astype(bfloat16)

# === Run NPU kernel ===
print("Running NPU kernel...")
elf_path = Path("air.elf")
assert elf_path.exists(), f"{elf_path} not found. Run make run --causal first."
artifact = XRTCompileArtifact(str(elf_path), "main:attention_bf16", None)
backend = XRTBackend(omit_while_true_loop=False, omit_pingpong="all",
    runtime_loop_tiling_sizes=[1, 1], output_format="elf", instance_name="attention_bf16")
out_buf = np.zeros((NUM_HEADS, LQ, DV), dtype=bfloat16)
with filelock.FileLock("/tmp/npu.lock"):
    invoker = backend.load(artifact)
    results = invoker(input_q_scaled, input_k, input_v, input_m, out_buf)
backend.unload()
npu_output = results[-1].reshape(NUM_HEADS, LQ, DV)

# === Compare ===
ref_f32 = ref_output.astype(np.float32).flatten()
npu_f32 = npu_output.astype(np.float32).flatten()
corr = np.corrcoef(npu_f32, ref_f32)[0, 1]
max_err = np.max(np.abs(npu_f32 - ref_f32))
print(f"\nNPU vs reference: corr={corr:.6f}, max_err={max_err:.4f}")

# === Tolerance sweep ===
print("\nTolerance sweep:")
for atol, rtol, label in [(0.5,0.2,"current"), (0.1,0.1,"moderate"),
                           (0.05,0.05,"reasonable"), (0.01,0.01,"strict")]:
    p = np.allclose(npu_f32, ref_f32, atol=atol, rtol=rtol)
    n = np.sum(np.abs(npu_f32-ref_f32) > atol+rtol*np.abs(ref_f32))
    print(f"  atol={atol}, rtol={rtol} ({label:10s})  {'PASS' if p else 'FAIL'}  ({100*n/len(ref_f32):.2f}% fail)")

# === Compare NPU output vs random noise ===
print("\nNPU output vs random noise (both checked with atol=0.5, rtol=0.2):")
print(f"  {'Scenario':35s}  {'Corr':>8s}  {'% fail':>8s}  {'% pass':>8s}")
for name, data in [
    ("NPU kernel output", npu_f32),
    ("Random uniform [same range]", rng.uniform(0.5, 1.0, len(ref_f32)).astype(np.float32)),
    ("Constant 0.75", np.full(len(ref_f32), 0.75, dtype=np.float32)),
]:
    c = np.corrcoef(data, ref_f32)[0, 1]
    nf = np.sum(np.abs(data - ref_f32) > 0.5 + 0.2*np.abs(ref_f32))
    cs = f"{c:.4f}" if not np.isnan(c) else "NaN"
    print(f"  {name:35s}  {cs:>8s}  {100*nf/len(ref_f32):7.1f}%  {100-100*nf/len(ref_f32):7.1f}%")
print("\nConclusion: NPU kernel is indistinguishable from random noise.")
print("Correlation is the true correctness metric (should be > 0.99).")
```

</details>

## Why the Current Test is Meaningless

The standalone test uses:
- **Input data**: `uniform(0, 1) * 0.5 + 0.5` → all values in `[0.5, 1.0]`
- **Tolerances**: `atol=0.5, rtol=0.2`

Attention output is `softmax(QK) @ V`, a weighted average of V rows. Since V ∈ [0.5, 1.0], the output must also be in [0.5, 1.0].

The `np.allclose` check accepts element `i` if: `|output[i] - ref[i]| <= atol + rtol * |ref[i]|`

For ref values in [0.5, 1.0]:
- Threshold = `0.5 + 0.2 * 0.5` to `0.5 + 0.2 * 1.0` = **[0.6, 0.7]**
- Any output in `[-0.2, 1.7]` passes

This means:

| "Output" | Correct? | Passes atol=0.5? |
|---|---|---|
| Correct attention output | Yes | Yes |
| Constant 0.75 everywhere | No | Yes |
| Random values in [0.5, 1.0] | No | Yes |
| All ones | No | Yes |
| All zeros | No | No (barely) |

**The test cannot distinguish correct output from garbage.** Only outputs outside [-0.2, 1.7] would be detected as wrong.

### Why "only 10% failing" does not mean "90% correct"

With wider-range inputs [0, 4] and `atol=0.5, rtol=0.2`, the NPU kernel shows 9.8% elements failing. This may appear to suggest the kernel is "90% correct". It is not. The table below compares the NPU kernel output against **pure random noise** — both checked against the reference:

**Narrow range [0.5, 1.0] (current test data):**
| Scenario | Corr | % fail | % pass |
|---|---|---|---|
| NPU kernel output | 0.086 | 0.0% | 100.0% |
| Random uniform [0.5, 1.0] | 0.000 | 0.0% | 100.0% |
| Constant 0.75 | NaN | 0.0% | 100.0% |

**Wide range [0, 4] (IRON-style inputs):**
| Scenario | Corr | % fail | % pass |
|---|---|---|---|
| NPU kernel output | 0.009 | 9.8% | 90.2% |
| Random normal [same distribution] | 0.000 | 7.3% | 92.7% |
| Constant (mean of reference) | NaN | 2.7% | 97.3% |

The NPU kernel's pass rate (90.2%) is **worse than random noise** (92.7%) and **worse than a constant** (97.3%). The element-wise pass % reflects the tolerance width relative to the data range, not correctness. **Correlation is the true metric**: a correct kernel should have corr > 0.99; this kernel has corr ≈ 0.01.

## Impact

This bug makes the causal flash attention kernel unusable for LLM inference:

- In LLAMA-3.2-1B 16-layer prefill, the attention error compounds across layers
- Per-kernel isolation test: flash attention has **corr=0.31** against standard attention; all other 14 kernels have **corr>0.999**
- 16-layer NPU output: completely wrong predictions (NPU: `','`, CPU: `' the'`)

### 1. Fix the test to actually validate correctness

Use tighter tolerances and/or wider-range test data:

```python
# Option A: Tighter tolerances (catches the bug)
runner.run_test(..., atol=0.05, rtol=0.05)

# Option B: Use standard normal data (wider range makes tolerances meaningful)
input_q = (rng.standard_normal((num_heads, lq, dk)) * 0.1).astype(bfloat16)
input_k = (rng.standard_normal((num_kv_heads, lk, dk)) * 0.1).astype(bfloat16)
input_v = (rng.standard_normal((num_kv_heads, lk, dv)) * 0.1).astype(bfloat16)

# Option C: Add correlation check (catches functional correctness)
corr = np.corrcoef(npu_output.flatten(), ref_output.flatten())[0, 1]
assert corr > 0.99, f"Output correlation {corr:.4f} is too low"
```

### 2. Fix the causal kernel correctness

The kernel needs to be debugged to produce output matching the `flash_attn_per_stage()` NumPy reference. The non-causal kernel may also have similar issues.

## Non-causal kernel has the same issue

Tested the non-causal kernel with the same methodology. Results are equally bad:

| | Non-causal | Causal |
|---|---|---|
| corr vs PyTorch ([0.5,1.0] data) | 0.127 | 0.086 |
| corr vs PyTorch ([0,4] data) | 0.014 | 0.009 |
| IRON tolerance, [0,4] data | FAIL (56.5% wrong) | FAIL (73.7% wrong) |
| Standalone `make run` | PASS! | PASS! |

**Both kernels are broken.** The issue is not specific to causal masking — it's fundamental to the kernel (likely DMA layout, cascade merge, or core computation).

## Comparison: IRON MHA test vs mlir-air flash attention test

For reference, the IRON MHA operator (`iron/operators/mha/`) uses significantly tighter validation:

| Aspect | IRON MHA (`test.py`) | mlir-air Flash Attention (`attn.py`) |
|--------|---------------------|-------------------------------------|
| **Input range** | `torch.rand() * 4` → [0, 4] | `uniform() * 0.5 + 0.5` → [0.5, 1.0] |
| **rel_tol** | **0.04** (4%) | **0.2** (20%) |
| **abs_tol** | **0.15** | **0.5** |
| **Reference** | PyTorch `scaled_dot_product_attention` | Own NumPy flash attention |
| **Causal** | `is_causal=True` | `--causal` flag |
| **Tolerance formula** | `diff < max(abs_tol, rel_tol * (|a|+|b|))` | `diff <= atol + rtol * |ref|` |
| **Error budget** | ≤0.5% elements may fail | ALL must pass np.allclose |

With IRON's wider input range [0, 4], the output spans a wider range too, so a 4% relative tolerance is meaningful — a wrong value is detected. With mlir-air's [0.5, 1.0] range and 20% rtol + 0.5 atol, the tolerance threshold (0.6–0.7) exceeds the entire output range (0.5), making the check vacuous.

### IRON-style validation applied to mlir-air flash attention

Tested the mlir-air causal kernel with IRON's methodology (inputs in [0, 4], PyTorch `scaled_dot_product_attention` as reference):

```
NPU vs PyTorch sdpa (causal, BF16):
  Correlation:    0.009  (essentially random noise)
  Max abs error:  3.84   (nearly the entire output range)
  Mean abs error: 0.42

IRON formula: diff < max(abs_tol, rel_tol * (|a|+|b|)):
  IRON default (rel=0.04, abs=0.15):   FAIL  73.70% elements wrong (budget: 0.5%)
  Relaxed 2.5x (rel=0.10, abs=0.30):   FAIL  42.48% elements wrong
  mlir-air vals (rel=0.20, abs=0.50):   FAIL  14.15% elements wrong

np.allclose:
  mlir-air current (atol=0.5, rtol=0.2):   FAIL  9.83% elements fail
  IRON-equivalent (atol=0.15, rtol=0.04):   FAIL  63.76% elements fail
```

With wider-range inputs, even the mlir-air's own loose tolerances fail. The original test only "passed" because [0.5, 1.0] data kept all errors within the huge tolerance window.

**Recommendation**: Adopt IRON-style validation:
- Use `torch.rand() * 4` for wider-range inputs
- Use `rel_tol=0.04, abs_tol=0.15` (IRON's values)
- Or use PyTorch's `scaled_dot_product_attention` as the reference (as IRON does)
